Adding PF-1.22: GUEv1 Decapsulation and ECMP test for IPv4 and IPv6 p…

keysightgems opened the pull request:
**ReadMe Location:** https://github.com/openconfig/featureprofiles/blob/main/feature/policy_forwarding/decapsulation/otg_tests/ipv4_guev1_decap_and_hashing_test/README.md

The log is attached here: https://partnerissuetracker.corp.google.com/issues/415458482

balaji6 commented on internal/cfgplugins/policyforwarding.go:
```diff
@@ -528,3 +528,37 @@ func MPLSStaticLSPConfig(t *testing.T, dut *ondatra.DUTDevice, ni *oc.NetworkIns
 	}
 }
+
+func ConfigureDutWithGueDecap(t *testing.T, dut *ondatra.DUTDevice, guePort int, ipType, tunIp, decapInt, policyName string, policyId int) {
```
Similar to https://github.com/openconfig/featureprofiles/pull/4385/files, can we reuse/modify the existing `func DecapGroupConfigGue` to achieve this?

nupkanoi commented on feature/policy_forwarding/decapsulation/otg_tests/ipv4_guev1_decap_and_hashing_test/ipv4_guev1_decap_and_hashing_test.go:
```diff
@@ -0,0 +1,1141 @@
+		jsonConfig := `
+		load-balance policies
+		load-balance sand profile default
+		packet-type gue outer-ip		
```
```suggestion
	 fields ipv6 outer dst-ip flow-label next-header src-ip
      fields l4 outer dst-port src-port
      no fields mpls
      packet-type gue outer-ip
```

keysightgems replied:
Changes done.

nupkanoi commented on feature/policy_forwarding/decapsulation/otg_tests/ipv4_guev1_decap_and_hashing_test/ipv4_guev1_decap_and_hashing_test.go:
```diff
@@ -0,0 +1,1141 @@
+		gnmiClient := dut.RawAPIs().GNMI(t)
+		jsonConfig := fmt.Sprintf(`		
+		router bgp %d
+		address-family ipv4 labeled-unicast
```
```suggestion
		address-family ipv4
```

nupkanoi commented on feature/policy_forwarding/decapsulation/otg_tests/ipv4_guev1_decap_and_hashing_test/ipv4_guev1_decap_and_hashing_test.go:
```diff
@@ -0,0 +1,1141 @@
+		address-family ipv4 labeled-unicast
+		maximum-paths %[2]d ecmp %[2]d
+		bgp bestpath as-path multipath-relax
+		address-family ipv6 labeled-unicast
```
```suggestion
		address-family ipv6
```

keysightgems replied:
updated

nupkanoi commented on feature/policy_forwarding/decapsulation/otg_tests/ipv4_guev1_decap_and_hashing_test/ipv4_guev1_decap_and_hashing_test.go:
```diff
@@ -0,0 +1,1141 @@
+		macAddress := gnmi.Get(t, dut, gnmi.OC().Interface(dut.Port(t, "port1").Name()).Ethernet().MacAddress().State())
+		otgConfig.Flows().Clear()
+		// Generate flows with Immediate next header's L3 fields
+		for flowIndex := 1; flowIndex <= 14; flowIndex++ {
```
```suggestion
		for flowIndex := 1; flowIndex <= 12; flowIndex++ {
```

I'll update the FNT as well

keysightgems replied:
ok, changes done

nupkanoi commented on feature/policy_forwarding/decapsulation/otg_tests/ipv4_guev1_decap_and_hashing_test/ipv4_guev1_decap_and_hashing_test.go:
```diff
@@ -0,0 +1,1141 @@
+		otgConfig.Flows().Clear()
+		// Generate flows with randomized L4 ports immediately after outer header
+		for flowIndex := 7; flowIndex <= 14; flowIndex++ {
+			configureFlows(t, otgConfig, macAddress, []string{otgConfig.Ports().Items()[1].Name()}, flowIndex, true)
```
The flows also need to have constant source and destination address for inner header, please https://github.com/openconfig/featureprofiles/blob/0dc67e16a744d4b40c4a6e88858f736ad0b80cae/feature/policy_forwarding/decapsulation/otg_tests/ipv4_guev1_decap_and_hashing_test/README.md#pf-1224-verify-the-immediate-next-headers-l4-fields-are-not-considered-in-load-balancing-algorithm

Refer below section:
- For IPv4 flows that would typically use an H1v4 range, use IPv4prefix1/24.
- For IPv6 flows that would typically use an H1v6 range, use IPv6prefix1/64.
- For IPv4 flows that would typically use an H2v4 range, use IPv4prefix11/24.
- For IPv6 flows that would typically use an H2v6 range, use IPv6prefix11/64.
- For IPv4 flows that would typically use an H4v4 range, use IPv4prefix31/24.
- For IPv6 flows that would typically use an H4v6 range, use IPv6prefix31/64.

keysightgems replied:
Missed as per the readme, updated now. Thanks!

nupkanoi commented on feature/policy_forwarding/decapsulation/otg_tests/ipv4_guev1_decap_and_hashing_test/ipv4_guev1_decap_and_hashing_test.go:
```diff
@@ -0,0 +1,1141 @@
+		otgConfig.Flows().Clear()
+		// Generate flows with Immediate next header's L3 fields
+		for flowIndex := 1; flowIndex <= 14; flowIndex++ {
+			configureFlows(t, otgConfig, macAddress, []string{otgConfig.Ports().Items()[1].Name()}, flowIndex, true)
```
The flows also need to have constant destination address for inner header, please refer below section from [here](https://github.com/openconfig/featureprofiles/blob/0dc67e16a744d4b40c4a6e88858f736ad0b80cae/feature/policy_forwarding/decapsulation/otg_tests/ipv4_guev1_decap_and_hashing_test/README.md#pf-1224-verify-the-immediate-next-headers-l4-fields-are-not-considered-in-load-balancing-algorithm):

- For IPv4 flows that would typically use an H1v4 range, use IPv4prefix1/24.
- For IPv6 flows that would typically use an H1v6 range, use IPv6prefix1/64.
- For IPv4 flows that would typically use an H2v4 range, use IPv4prefix11/24.
- For IPv6 flows that would typically use an H2v6 range, use IPv6prefix11/64.
- For IPv4 flows that would typically use an H4v4 range, use IPv4prefix31/24.
- For IPv6 flows that would typically use an H4v6 range, use IPv6prefix31/64.

keysightgems replied:
Missed as per the readme, updated now. Thanks!

nupkanoi commented on feature/policy_forwarding/decapsulation/otg_tests/ipv4_guev1_decap_and_hashing_test/ipv4_guev1_decap_and_hashing_test.go:
```diff
@@ -0,0 +1,1141 @@
+			ipInner.Dst().SetValue(host4IPv6Start)
+		}
+		udp := flow.Packet().Add().Udp()
+		udp.SrcPort().Increment().SetStart(UdpSrcPort - 1).SetStep(1).SetCount(10)
```
Why are we incrementing/randomizing only this flow source port ? The randomizing is supposed to happen for all flows exception flows 11, 12 , 13 and 14

keysightgems commented on feature/policy_forwarding/decapsulation/otg_tests/ipv4_guev1_decap_and_hashing_test/ipv4_guev1_decap_and_hashing_test.go:
```diff
@@ -0,0 +1,1141 @@
+		macAddress := gnmi.Get(t, dut, gnmi.OC().Interface(dut.Port(t, "port1").Name()).Ethernet().MacAddress().State())
+		otgConfig.Flows().Clear()
+		// Generate flows with randomized L4 ports immediately after outer header
+		for flowIndex := 7; flowIndex <= 14; flowIndex++ {
```
ok, changes done.

keysightgems commented on feature/policy_forwarding/decapsulation/otg_tests/ipv4_guev1_decap_and_hashing_test/ipv4_guev1_decap_and_hashing_test.go:
```diff
@@ -0,0 +1,1151 @@
+		{IPv4: ateP7.IPv4, IPv6: ateP7.IPv6},
+	}
+	checkBgpStatus(t, dut, neighbors)
+	t.Run("PF-1.22.1[Baseline]: GUE Decapsulation over ipv4 decap address and Load-balance test", func(t *testing.T) {
```
You're right, the base config we only set the host[1-4]IPv[4,6]Start addresses. The expansion into multiple prefixes happens inside the traffic generation code for sub-tests 1, 2, and 3, where those start addresses are used as seeds. From there, we derive ranges (e.g. incremented /24s for IPv4, /64s for IPv6) and assign them to the flows so that each host actually contributes the expected number of prefixes.

That’s why you don’t see 10 explicit /24 or /64 entries in the static config — the logic is implemented programmatically at flow creation time.

nupkanoi replied:
ack. SG

dplore commented:
After providing changes, please provide/update the logs of the test execution.

keysightgems commented on internal/cfgplugins/policyforwarding.go:
```diff
@@ -516,6 +524,20 @@ func DecapGroupConfigGue(t *testing.T, dut *ondatra.DUTDevice, pf *oc.NetworkIns
 	}
 }
 
+func GueDecapConfig(t *testing.T, dut *ondatra.DUTDevice, params OcPolicyForwardingParams) {
```
updated as per the comments. Thanks!

keysightgems commented on feature/policy_forwarding/decapsulation/otg_tests/ipv4_guev1_decap_and_hashing_test/ipv4_guev1_decap_and_hashing_test.go:
```diff
@@ -0,0 +1,1151 @@
+
+	// Apply the whole BGP config
+	gnmi.Replace(t, dut, dutBgpConfPath.Config(), dutBgpConf)
+	if deviations.LoadIntervalNotSupported(dut) {
+		gnmiClient := dut.RawAPIs().GNMI(t)
+		jsonConfig := `
+		load-balance policies
+		load-balance sand profile default
+		fields ipv6 outer dst-ip flow-label next-header src-ip
+        fields l4 outer dst-port src-port
+        no fields mpls
+        packet-type gue outer-ip
+		`
+		gpbSetRequest := buildCliConfigRequest(jsonConfig)
+
+		if _, err := gnmiClient.Set(context.Background(), gpbSetRequest); err != nil {
+			t.Fatalf("gnmiClient.Set() with unexpected error: %v", err)
+		}
+	} else {
+		// TODO: OC does not yet support selecting the load-balancing hash mode on LAG members.
+		t.Logf("Load balancing is currently not supported via OpenConfig. Will fix once it's implemented.")
+	}
```
Done changes and moved to suggested file.

keysightgems commented on feature/policy_forwarding/decapsulation/otg_tests/ipv4_guev1_decap_and_hashing_test/ipv4_guev1_decap_and_hashing_test.go:
```diff
@@ -0,0 +1,1151 @@
+	gnmi.Update(t, dut, gnmi.OC().Config(), d)
+}
+
+func clearLAGInterfaces(t *testing.T, dut *ondatra.DUTDevice, aggPorts []*ondatra.Port, aggID string) {
```
Updated!

keysightgems commented on feature/policy_forwarding/decapsulation/otg_tests/ipv4_guev1_decap_and_hashing_test/ipv4_guev1_decap_and_hashing_test.go:
```diff
@@ -0,0 +1,1151 @@
+
+	dutBgpConfPath := d.NetworkInstance(deviations.DefaultNetworkInstance(dut)).Protocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_BGP, "BGP")
+
+	// Create BGP config only once
+	dutBgpConf := &oc.NetworkInstance_Protocol{
+		Identifier: oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_BGP,
+		Name:       ygot.String("BGP"),
+		Bgp:        &oc.NetworkInstance_Protocol_Bgp{},
+	}
+
+	bgp := dutBgpConf.Bgp
+	global := bgp.GetOrCreateGlobal()
+	global.As = ygot.Uint32(dutAS)
+	global.RouterId = ygot.String(dutP1.IPv4)
+
+	af4 := global.GetOrCreateAfiSafi(oc.BgpTypes_AFI_SAFI_TYPE_IPV4_UNICAST)
+	af4.Enabled = ygot.Bool(true)
+	af6 := global.GetOrCreateAfiSafi(oc.BgpTypes_AFI_SAFI_TYPE_IPV6_UNICAST)
+	af6.Enabled = ygot.Bool(true)
+
+	// Append EBGP Neighbors - ATE1 and ATE5
+	appendBGPNeighbor(t, bgp, ate1AS, dutP1.Name, ateP1.IPv4, ateP1.IPv6, false)
+	appendBGPNeighbor(t, bgp, ate5AS, dutP7.Name, ateP7.IPv4, ateP7.IPv6, false)
+	// Append IBGP Neighbor - ATE2
+	appendBGPNeighbor(t, bgp, ate2AS, dutP2.Name, ateP2.IPv4, ateP2.IPv6, false)
+	appendBGPNeighbor(t, bgp, ate3AS, dutLag1.Name, ateLag1.IPv4, ateLag1.IPv6, true)
+	appendBGPNeighbor(t, bgp, ate4AS, dutLag2.Name, ateLag2.IPv4, ateLag2.IPv6, true)
```
updated as per the comment!

keysightgems commented on feature/policy_forwarding/decapsulation/otg_tests/ipv4_guev1_decap_and_hashing_test/ipv4_guev1_decap_and_hashing_test.go:
```diff
@@ -0,0 +1,1151 @@
+			t.Fatalf("gnmiClient.Set() with unexpected error: %v", err)
+		}
+	} else {
+		// TODO: As per the latest OpenConfig GNMI OC schema — the Encapsulation/Decapsulation sub-tree is not fully implemented, need to add OC commands once implemented.
```
Yes correct, updated the comment.

keysightgems commented:
> After providing changes, please provide/update the logs of the test execution.

updated latest logs here https://partnerissuetracker.corp.google.com/u/1/issues/415458482

keysightgems commented on feature/policy_forwarding/decapsulation/otg_tests/ipv4_guev1_decap_and_hashing_test/ipv4_guev1_decap_and_hashing_test.go:
```diff
@@ -0,0 +1,980 @@
+		gnmi.Delete(t, dut, gnmi.OC().Interface(port.Name()).Ethernet().Config())
+	}
+	cfgplugins.SetupStaticAggregateAtomically(t, dut, aggPorts, aggID)
+	agg := dutLag.NewOCInterface(aggID, dut)
+	agg.Type = oc.IETFInterfaces_InterfaceType_ieee8023adLag
+	agg.GetOrCreateAggregation().LagType = oc.IfAggregate_AggregationType_STATIC
+	gnmi.Replace(t, dut, gnmi.OC().Interface(aggID).Config(), agg)
+	for _, port := range aggPorts {
+		d := &oc.Root{}
+		i := d.GetOrCreateInterface(port.Name())
+		i.GetOrCreateEthernet().AggregateId = ygot.String(aggID)
+		i.Type = oc.IETFInterfaces_InterfaceType_ethernetCsmacd
+		if deviations.InterfaceEnabled(dut) {
+			i.Enabled = ygot.Bool(true)
+		}
+		gnmi.Replace(t, dut, gnmi.OC().Interface(port.Name()).Config(), i)
+	}
```
I agree there is some overlap, but configureDUTLag and SetupStaticAggregateAtomically are serving different purposes:

**SetupStaticAggregateAtomically :** Creates the aggregate interface, member interfaces in one atomic GNMI update.

**configureDUTLag** is the public API for creating a static LAG on the DUT. It wraps SetupStaticAggregateAtomically to ensure atomic creation and applies additional DUT-specific attributes and cleanup.

Getting lag ports not available error when removed the duplicate code.

**Script execution log :** failed to apply: Ethernet4/1: cannot add aggregate id Port-Channel1, lag-type is not configured

Took reference from existing script: 
https://github.com/openconfig/featureprofiles/blob/main/feature/gribi/otg_tests/ipv4_entry_with_aggregate_ports_test/ipv4_entry_with_aggregate_ports_test.go

dplore replied:
It sounds like we don't understand why this code is working. The call to  SetupStaticAggregateAtomically [on line 410](https://github.com/openconfig/featureprofiles/pull/4375/files#diff-08b6d54e4485d6898e5058c0a6bd7f9e0de4aa0de036ea796a732e30279acfeeR410) already does the exact same OC as [the code below it.](https://github.com/openconfig/featureprofiles/pull/4375/files#diff-08b6d54e4485d6898e5058c0a6bd7f9e0de4aa0de036ea796a732e30279acfeeR411-R422
)  

Please resolve this so the calls are not duplicated and instead are all in the cfgplugin.  

The reason I'm being so picky on this is I am wondering if some unknown issue is being hidden by applying these configurations to the device twice.  We would not apply interface configuration twice in a production use case.

keysightgems commented:
> @keysightgems please address comments so we can complete soon.

Addressed all the comments and updated the code.
please find the updated latest logs here https://partnerissuetracker.corp.google.com/u/1/issues/415458482.

dplore commented on internal/cfgplugins/interface.go:
```diff
@@ -725,3 +725,23 @@ func DeleteAggregate(t *testing.T, dut *ondatra.DUTDevice, aggID string, dutAggP
+			i.Enabled = ygot.Bool(true)
+		}
+	}
+	gnmi.Update(t, dut, gnmi.OC().Config(), d)
```
this function should include a gnmi Batch in the parameters and return the same.  See this example for the style:
https://github.com/openconfig/featureprofiles/blob/bc105a443b44d862c70e91112708b5a339c71ae5/internal/cfgplugins/sflow.go#L33-L55

dplore commented on internal/cfgplugins/isis.go:
```diff
@@ -0,0 +1,64 @@
+
+// ConfigureISISOnDUT configures ISIS on the DUT using OpenConfig.
+// It enables ISIS globally, sets AFs, and applies interface-level config.
+func ConfigureISISOnDUT(t *testing.T, dut *ondatra.DUTDevice, cfg ISISConfig) *oc.NetworkInstance_Protocol {
```
```suggestion
func NewISIS(t *testing.T, dut *ondatra.DUTDevice, cfg ISISConfig) *oc.NetworkInstance_Protocol {
```

dplore replied:
This function should return a gnmi batch object.  See https://github.com/openconfig/featureprofiles/blob/main/internal/cfgplugins/sflow.go

dplore commented on internal/cfgplugins/interface.go:
```diff
@@ -725,3 +726,28 @@ func DeleteAggregate(t *testing.T, dut *ondatra.DUTDevice, aggID string, dutAggP
 }
+
+// SetupStaticAggregateAtomically sets up the static aggregate interface atomically.
+func SetupStaticAggregateAtomically(t *testing.T, dut *ondatra.DUTDevice, aggPorts []*ondatra.Port, aggID string, dutLag attrs.Attributes) {
```
The interface config parameters should be in a struct.  For example:
https://github.com/openconfig/featureprofiles/blob/7ce57f30158e82e164f0322ec34c3cd353dad773/internal/cfgplugins/sflow.go#L43

See the [README for cfgplugins](https://github.com/openconfig/featureprofiles/blob/main/internal/cfgplugins/README.md)

dplore commented on internal/cfgplugins/interface.go:
```diff
@@ -725,3 +726,28 @@ func DeleteAggregate(t *testing.T, dut *ondatra.DUTDevice, aggID string, dutAggP
+		gnmi.BatchReplace(batch, gnmi.OC().Interface(port.Name()).Config(), i)
+	}
+
+	batch.Set(t, dut)
```
batch.Set should be in the _test.go file

dplore commented on internal/cfgplugins/bgp.go:
```diff
@@ -622,3 +624,112 @@ func ConfigureBGPNeighbor(t *testing.T, dut *ondatra.DUTDevice, ni *oc.NetworkIn
 }
+
+// ConfigureDUTBGP configures BGP on the DUT using OpenConfig.
+func ConfigureDUTBGP(t *testing.T, dut *ondatra.DUTDevice, dutAS, ate1AS, ate2AS, ate3AS, ate4AS, ate5AS, ecmpMaxPath uint32,
```
This function should probably only configure the Global BGP parameters.

The appendBGPNeighbor should be exported and called by the _test.go.  This makes the code more reusable (so a test can decide how many neighbors are needed)

dplore commented on internal/cfgplugins/bgp.go:
```diff
@@ -622,3 +624,112 @@ func ConfigureBGPNeighbor(t *testing.T, dut *ondatra.DUTDevice, ni *oc.NetworkIn
+	af6.Enabled = ygot.Bool(true)
+
+	// Append BGP neighbors
+	appendBGPNeighbor(t, bgp, ate1AS, dutP1.Name, ateP1.IPv4, ateP1.IPv6, false)
+	appendBGPNeighbor(t, bgp, ate5AS, dutP7.Name, ateP7.IPv4, ateP7.IPv6, false)
+	appendBGPNeighbor(t, bgp, ate2AS, dutP2.Name, ateP2.IPv4, ateP2.IPv6, false)
+	appendBGPNeighbor(t, bgp, ate3AS, dutLag1.Name, ateLag1.IPv4, ateLag1.IPv6, true)
+	appendBGPNeighbor(t, bgp, ate4AS, dutLag2.Name, ateLag2.IPv4, ateLag2.IPv6, true)
```
ad noted above, these should be in the _test.go file

dplore commented on internal/cfgplugins/bgp.go:
```diff
@@ -622,3 +624,112 @@ func ConfigureBGPNeighbor(t *testing.T, dut *ondatra.DUTDevice, ni *oc.NetworkIn
+}
+
+// appendBGPNeighbor appends IPv4 and IPv6 neighbors (and their peer groups) to the BGP configuration.
+func appendBGPNeighbor(t *testing.T, bgp *oc.NetworkInstance_Protocol_Bgp, ateAs uint32, portName, neighborIpV4, neighborIpV6 string, isLag bool) {
```
This should be exported (ie: capitalized) so _test.go files can import and call this function.

As with other exported functions, the parameters should be a struct

dplore commented:
Please also provide fresh test results

keysightgems commented on proto/metadata.proto:
```diff
@@ -938,8 +941,12 @@ message Metadata {
     bool qos_two_rate_three_color_policer_oc_unsupported = 325;
 
+    // Devices that do not support load balance policies 
+    // Arista b/445043741
```
Updated with URL.

keysightgems commented on proto/metadata.proto:
```diff
@@ -405,6 +405,9 @@ message Metadata {
     bool skip_isis_set_metric_style_type = 145;
+    // Skip setting match-prefix-set match-set-options while configuring
+    // routing-policy statement condition
+    bool skip_set_rp_match_set_options = 146;
```
**skip_set_rp_match_set_options** is not part of my code.

dplore replied:
ah, then your PR doesn't need to add this deviation?

dplore replied:
Looks like it was removed in #4484

dplore replied:
rebase your fork to the main branch should fix this?

keysightgems replied:
Fixed, Please Approve.
